WordPress/src/main/java/org/wordpress/android/ui/jetpack/backup/download/BackupDownloadViewModel.kt

```
@@ -131,6 +161,180 @@ class BackupDownloadViewModel @Inject constructor(
+        val items = _uiState.value?.items ?: mutableListOf()
+        val types = buildBackupDownloadRequestTypes(items)
+        return rewindId to types
+    }
+
+    private fun buildBackupDownloadRequestTypes(items: List<JetpackListItemState>): BackupDownloadRequestTypes {
+        val checkboxes = items.filterIsInstance(CheckboxState::class.java)
+
+        return BackupDownloadRequestTypes(
+                themes = checkboxes.firstOrNull { it.availableItemType == THEMES }?.checked ?: true,
+                plugins = checkboxes.firstOrNull { it.availableItemType == PLUGINS }?.checked
+                        ?: true,
+                uploads = checkboxes.firstOrNull { it.availableItemType == MEDIA_UPLOADS }?.checked
+                        ?: true,
+                sqls = checkboxes.firstOrNull { it.availableItemType == SQLS }?.checked ?: true,
+                roots = checkboxes.firstOrNull { it.availableItemType == ROOTS }?.checked ?: true,
+                contents = checkboxes.firstOrNull { it.availableItemType == CONTENTS }?.checked
+                        ?: true
+        )
+    }
+
+    private fun handleBackupDownloadRequestResult(result: BackupDownloadRequestState) {
+        when (result) {
+            is NetworkUnavailable -> {
+                _snackbarEvents.postValue(Event(NetworkUnavailableMsg))
+            }
+            is RemoteRequestFailure -> {
+                _snackbarEvents.postValue(Event(GenericFailureMsg))
+            }
+            is Success -> {
+                backupDownloadState = backupDownloadState.copy(
+                        rewindId = result.rewindId,
+                        downloadId = result.downloadId,
+                        published = extractPublishedDate())
+                wizardManager.showNextStep()
+            }
+            is OtherRequestRunning -> {
+                _snackbarEvents.postValue(Event(OtherRequestRunningMsg))
+            }
+            else -> {
```

Feel free to ignore, it's not related to this PR.

🔍 Nitpick: If this is not an expected state, I'd consider throwing an exception (fail fast principle). Wdyt?

---

> If this is not an expected state, I'd consider throwing an exception (fail fast principle). Wdyt?

This is an edge case handler, we shouldn't fail if the app is left open on this view and someone clicks a different backup on the web. We show snackbar that there is already a request in progress. This is part of the unhappy state.

---

ohh, I see the `when` doesn't not include all branches and some of them are reachable in edge cases. Thanks for the clarification!